grant_disbursement/src/main.leo

```diff
@@ -0,0 +1,233 @@
+    return finalize_initialize();
+  }
+
+  async function finalize_initialize() {
+    // Ensure the protocol is not already initialized
+    let already_initialized: bool = grants.contains(0u8);
+    assert(!already_initialized);
+
+    // Create the initial grant
+    let initial_grant: Grant = Grant {
+      credits_amount: CREDITS_AMOUNT,
+      recipient_rewards_address: GRANTEE_REWARDS_ADDRESS,
+      recipient_principal_address: GRANTEE_PRINCIPAL_ADDRESS,
+      start_timestamp: START_TIMESTAMP,
+      cliff_timestamp: CLIFF_TIMESTAMP,
+      fully_vested_timestamp: FULLY_VESTED_TIMESTAMP,
+      is_processed: false
+    };
+
+    // Add the grant to the mapping
+    grants.set(0u8, initial_grant);
+  }
+
+  async transition process_grant(
+    grant_id: u8,
+    credits_amount: u64,
+  ) -> Future {
+    // transfer credits to the protocol
+    let f0: Future = credits.aleo/transfer_public_as_signer(self.address, credits_amount);
+    // bond the credits to the puzzle validator
+    let f1: Future = credits.aleo/bond_public(PUZZLE_VALIDATOR_ADDRESS, self.address, credits_amount);
+
+    return finalize_process_grant(f0, f1, grant_id, credits_amount);
+  }
+
+  async function finalize_process_grant(
+    f0: Future,
+    f1: Future,
+    grant_id: u8,
+    credits_amount: u64,
+  ) {
+    f0.await();
+    f1.await();
+
+    // Get the grant
+    let grant: Grant = grants.get(grant_id);
+
+    // Assert the credits amount is the same
+    assert_eq(credits_amount, grant.credits_amount);
+    // Assert that the grant has not been processed
+    assert(!grant.is_processed);
+
+    // Update the grant
+    let updated_grant: Grant = Grant {
+      credits_amount: grant.credits_amount,
+      recipient_rewards_address: grant.recipient_rewards_address,
+      recipient_principal_address: grant.recipient_principal_address,
+      start_timestamp: grant.start_timestamp,
+      cliff_timestamp: grant.cliff_timestamp,
+      fully_vested_timestamp: grant.fully_vested_timestamp,
+      is_processed: true
+    };
+    grants.set(grant_id, updated_grant);
+  }
+
+  async transition withdraw_rewards(
+    id: u8,
+    reward_amount: u64,
+  ) -> Future {
+    // unbond an arbitrary number of credits from the program
+    let f0: Future = credits.aleo/unbond_public(self.address, reward_amount);
+
+    return finalize_withdraw_rewards(f0, id, reward_amount, self.caller);
+  }
+
+  async function finalize_withdraw_rewards(
+    f0: Future,
+    id: u8,
+    reward_amount: u64,
+    caller: address
+  ) {
+    f0.await();
+
+    // Get the grant
+    let grant: Grant = grants.get(id);
+
+    // Ensure the caller is the recipient_rewards_address
+    assert_eq(caller, grant.recipient_rewards_address);
+
+    // Get the current timestamp
+    let current_timestamp: u64 = time_oracle.aleo/timestamp.get(0u8);
+
+    // Assert that the current timestamp is before the cliff
+    assert(current_timestamp < grant.cliff_timestamp);
```

Comment:
i'm not sure why this assertion is necessary -- what if we want to claim rewards after the cliff? we would just use the `withdraw_principal` function?